### PR TITLE
TEST: 1.21.x + blas variants

### PR DESCRIPTION
**Pull request by h-vetinari**
Continuing the analysis from #227 & #196. Should not be merged for the same reasons as #227.

### CODE REVIEW

**Comment by h-vetinari** [comment]
### Update for 1.21.0

From 1 failure out of 64 for [1.20.3](https://github.com/conda-forge/numpy-feedstock/pull/227#issuecomment-870867638), there are now 4 (mostly flaky) failures.

_Note_: travis seems to be hanging across the board, but should pass once restarted.. Same expectation for aarch, which is still stuck in a long queue. Will update this comment if necessary.

#### The badnews:
* win+blis remains flaky
* broken pipes reappeared on win

### Details

| lib | [before](https://github.com/conda-forge/numpy-feedstock/pull/227#issuecomment-870867638) | after | updated |
|--------|-------|----|-|
| `numpy` | `1.20.3` | `1.21.0` | X |
| `libblas` | `3.9.0-9` | `3.9.0-9` | |
| `blis` | `0.8.1-0` | `0.8.1-0` | |
| `openblas` | `0.3.15-pthreads-1` | `0.3.15-pthreads-1` | |
| `mkl` | `2021.2-389` | `2021.2-389` | |
| `netlib` | `3.9.0-5` | `3.9.0-5` | |
| `pypy` | `7.3.4-4` | `7.3.4-4` | |

----------

| variant | [before](https://github.com/conda-forge/numpy-feedstock/pull/227#issuecomment-870867638) | after |
|--------|-------|----|
| win + blis | 12 failures for py37-only | 12 failures for py39-only |
| win | Passed | Reappearing failures due to `The process tried to write to a nonexistent pipe.` |

---------------

| variant| blis | mkl | netlib | openblas | sum* | 
|-----|------|----|----|----|----|
| linux / x86 | ✔️| ✔️ | ✔️ | ✔️ | - | 
| linux / aarch | ➖ | ➖ | ✔️ | ✔️ | - | 
| linux / ppc64le | ➖ | ➖ | ✔️ | ✔️ | - |
| osx / arm | ➖ | ➖ | ✔️ | ✔️ | - |
| osx / x86 | ✔️ | ✔️ | ✔️ | ✔️ | - |
| win / x86 | ✔️ / ❌ | ✔️ / ❌ | ✔️ / ❌ | ✔️ / ❌ | 4F |
| sum* | 1F | 1F | 1F | 1F | 4F |

`*` sum of Failures (out of a total of 64 CI combinations being tested)

Build logs:
[Azure](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=344249&view=results) 
[Drone](https://cloud.drone.io/conda-forge/numpy-feedstock/241)
[Travis](https://travis-ci.com/github/conda-forge/numpy-feedstock/builds/231281549) 


<details>
<summary>win + blis + cpython 3.9: 12 failures</summary>

```
=========================== short test summary info ===========================
FAILED core/tests/test_multiarray.py::TestMatmul::test_dot_equivalent[args4]
FAILED core/tests/test_multiarray.py::TestMatmul::test_matmul_object - Assert...
FAILED linalg/tests/test_linalg.py::TestSolve::test_sq_cases - AssertionError...
FAILED linalg/tests/test_linalg.py::TestSolve::test_generalized_sq_cases - As...
FAILED linalg/tests/test_linalg.py::TestInv::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestInv::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_sq_cases - Ass...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_nonsq_cases - ...
FAILED linalg/tests/test_linalg.py::TestDet::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestDet::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestMatrixPower::test_power_is_minus_one[dt13]
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - Asser...
= 12 failed, 16017 passed, 354 skipped, 20 xfailed, 1 xpassed, 229 warnings in 622.85s (0:10:22) =
```

</details>

**Comment by h-vetinari** [comment]
Finally opened an issue for blis: https://github.com/flame/blis/issues/514

**Comment by h-vetinari** [comment]
So I had to restart the CI because travis died (and I don't have restart rights). This now lead to the reappearance of https://github.com/numpy/numpy/issues/19192, though exclusively for PyPy.

@mattip @r-devulap, is it possible that something is messing with the glibc-detection used in https://github.com/numpy/numpy/pull/19209 on PyPy? Also, I really don't understand why this passed an hour before (with the exact same commit).

**Comment by mattip** [comment]
> I really don't understand why this passed an hour before

It may be run on different machines, some with AVX512 some without

**Comment by mattip** [comment]
It seems NumPy master has extended `numpy.show_config` to show which CPU features are detected. It would be nice if we could see that here, it would allow reasoning about runs on different CI machines

File: recipe/meta.yaml [hunk 1]
```diff
@@ -64,6 +64,8 @@ test:
     - blas =*=*{{ blas_impl }}
   commands:
     - f2py -h
+    # show SIMD features; change to np.show_config() for v>=1.22.0
+    - python -c 'import numpy; numpy._pytesttester._show_numpy_info()'
```
Comment (mattip) [inline recipe/meta.yaml:68]: I think windows [chokes](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=344546&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=1183ba29-a0b5-5324-8463-2a49ace9e213&l=3792) on single quotes?

Comment (h-vetinari) [reply recipe/meta.yaml:68]: Ehgads this is annoying. I copied from the [output](https://github.com/conda-forge/numpy-feedstock/pull/235#issuecomment-849177266) not the [source](https://github.com/conda-forge/numpy-feedstock/pull/235#issuecomment-849167449)

**Comment by h-vetinari** [comment]
> 
> 
> So I had to restart the CI because travis died (and I don't have restart rights). This now lead to the reappearance of [numpy/numpy#19192](https://github.com/numpy/numpy/issues/19192), though exclusively for PyPy.
> 
> @mattip @r-devulap, is it possible that something is messing with the glibc-detection used in [numpy/numpy#19209](https://github.com/numpy/numpy/pull/19209) on PyPy? Also, I really don't understand why this passed an hour before (with the exact same commit).

After following @mattip's tip for investigating the SIMD capabilities of the agents again, we're basically reconfirming numpy/numpy#19192 (failing runs have `AVX512F? AVX512CD? AVX512_SKX?`, passing runs have `AVX512F* AVX512CD* AVX512_SKX*`), but that just shows that the glibc-version-skip introduced in https://github.com/numpy/numpy/pull/19209 does not properly work on PyPy for some reason.

**Comment by h-vetinari** [comment]
> 
> 
> Finally opened an issue for blis: [flame/blis#514](https://github.com/flame/blis/issues/514)

However, the SIMD check _did_ help to verify that the blis failures are not actually flaky, but happen in the _presence_ of AVX512.

**Comment by h-vetinari** [comment]
> [...] that just shows that the glibc-version-skip introduced in [numpy/numpy#19209](https://github.com/numpy/numpy/pull/19209) does not properly work on PyPy for some reason.

@mattip, is `ver = os.confstr('CS_GNU_LIBC_VERSION').rsplit(' ')[1]` supposed to work on PyPy? Or perhaps, what would be the correct way to pick up the system glibc version?

**Comment by r-devulap** [comment]
Hi @h-vetinari looks like `ver = os.confstr('CS_GNU_LIBC_VERSION').rsplit(' ')[1]` doesn't work on pypy3. It throws an `ValueError: unrecognized configuration name`. This explains why you are seeing the error again. I am not sure how to make it work on pypy3, still looking into it..

**Comment by h-vetinari** [comment]
> This explains why you are seeing the error again. I am not sure how to make it work on pypy3, still looking into it..

Thanks a lot for investigating!

**Comment by mattip** [comment]
PyPy does not implement that value. I opened [an issue](https://foss.heptapod.net/pypy/pypy/-/issues/3502) in PyPy and a corresponding one in NumPy numpy/numpy#19385. Note that packaging.tags uses a [ctypes workaround](https://github.com/pypa/packaging/blob/main/packaging/tags.py#L496), but that seems like overkill for a problem that PyPy should solve.

**Comment by h-vetinari** [comment]
> 
> 
> PyPy does not implement that value. I opened [an issue](https://foss.heptapod.net/pypy/pypy/-/issues/3502) in PyPy and a corresponding one in NumPy [numpy/numpy#19385](https://github.com/numpy/numpy/issues/19385).

Thanks! :)

**Comment by mattip** [comment]
PyPy issue is fixed. Is it worth making a patch and releasing a new pypy7.3.5 build? The next PyPy release will probably be a few months coming, and I don't know how common it is to use PyPy + centos7

**Comment by h-vetinari** [comment]
> 
> 
> PyPy issue is fixed. Is it worth making a patch and releasing a new pypy7.3.5 build?

I think that would be worthwhile, carrying a patch is not a big deal IMO.

> The next PyPy release will probably be a few months coming, and I don't know how common it is to use PyPy + centos7

Maybe I misunderstand, but CentOS 6/7 are just stand-ins for linux here, where PyPy usage is highest.

**Comment by mattip** [comment]
I mis-stated the failing combination above. It is PyPy + glibc2.12, which was found on centos6, not centos7. Centos6 is EOL since Nov 2020. But for some reason the conda environment [uses it](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=344559&view=logs&j=517fe804-fa30-5dc2-1413-330699242c05&t=c10fa5f2-fdf6-5338-3bdb-c4bea7c23412&l=178).

**Comment by h-vetinari** [comment]
> Centos6 is EOL since Nov 2020. But for some reason the conda environment [uses it](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=344559&view=logs&j=517fe804-fa30-5dc2-1413-330699242c05&t=c10fa5f2-fdf6-5338-3bdb-c4bea7c23412&l=178).

See here: https://github.com/conda-forge/conda-forge.github.io/issues/1436

**Comment by isuruf** [comment]
>  But for some reason the conda environment uses it.

It's the same reason that numpy supports manylinux2010 (which is glibc 2.12). :wink:

**Comment by mattip** [comment]
NumPy uses manylinux2010 not to support the outdated CentOS6, but because it still supports older linux versions that may not have pip v20. I am not sure conda has the same problem.

**Comment by h-vetinari** [comment]
> I am not sure conda has the same problem.

A very similar one - once conda moves off of CentOS 6, the packages built for linux are not usable on older distros anymore. As can be seen from the issue I [linked](https://github.com/conda-forge/conda-forge.github.io/issues/1436), a move away from this is on the horizon, but that wasn't realistic or desirable until quite recently.

**Comment by h-vetinari** [comment]
## First completely green build! 🚀 🥳 🎉

### Updates galore, numpy 1.21.1, new blas builds, pypy on windows

* A lot has happened since the last update; the problems with blis were determined to be related to the AVX512, and got resolved with https://github.com/flame/blis/issues/514
* With https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/158, there should now be no more windows failures due to broken pipes.

From 4 failure out of 64 for 1.21.0, we're down to zero failures (out of 76; 4 new win+pypy builds; +8 extra blis builds to test presence/absence of avx512) for the first time in this series. 🥳

### Details

| lib | [before](https://github.com/conda-forge/numpy-feedstock/pull/237#issuecomment-870877267) | after | updated<br/>version | updated<br/>build |
|--------|-------|----|-|-|
| `numpy` | `1.21.0` | `1.21.1` | X | |
| `libblas` | `3.9.0-9` | `3.9.0-10` | | X |
| `blis` | `0.8.1-0` | `0.8.1-1` | | X |
| `openblas` | `0.3.15-pthreads-1` | `0.3.17-pthreads-1` | X | |
| `mkl` | `2021.2-389` | `2021.3.0-557` | X | |
| `netlib` | `3.9.0-5` | `3.9.0-5` | | |
| `pypy` | `7.3.4-4` | `7.3.5-7` | X | |

----------

| variant | [before](https://github.com/conda-forge/numpy-feedstock/pull/237#issuecomment-870877267) | after |
|--------|-------|----|
| win + blis | 12 failures with avx512 | passes 🥳 |
| win | Reappearing failures due to `The process tried to write to a nonexistent pipe.` | gone 🥳 |

---------------

| variant| blis<br/>w/ avx512 | blis<br/>w/o avx512 | mkl | netlib | openblas | sum* | 
|-----|------|----|----|----|----|---|
| linux / x86 | ✔️| ✔️ | ✔️ | ✔️ | ✔️ | - | 
| linux / aarch | ➖ | ➖ | ➖ | ✔️ | ✔️ | - | 
| linux / ppc64le | ➖ | ➖ | ➖ | ✔️ | ✔️ | - |
| osx / arm | ➖ | ➖ | ➖ | ✔️ | ✔️ | - |
| osx / x86 | ➖ | ✔️ | ✔️ | ✔️ | ✔️ | - |
| win / x86 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | - |
| sum* | - | - | - | - | - | - |

`*` sum of Failures (out of a total of 76 CI combinations being tested)

Build logs:
[Azure](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=361415&view=results) 
[Drone](https://cloud.drone.io/conda-forge/numpy-feedstock/274)
[Travis](https://app.travis-ci.com/github/conda-forge/numpy-feedstock/builds/234765739)

**Comment by mattip** [comment]
Whohoo!

File: recipe/meta.yaml [hunk 2]
```diff
@@ -67,6 +67,8 @@ test:
     - blas =*=*{{ blas_impl }}
   commands:
     - f2py -h
+    - python -c "import sys; import numpy; sys.exit(0 if 'AVX512F? AVX512CD?' in numpy.lib.utils._opt_info() else 1)"  # [enforce_avx512 == "present" and blas_impl == "blis"]
+    - python -c "import sys; import numpy; sys.exit(0 if 'AVX512F* AVX512CD*' in numpy.lib.utils._opt_info() else 1)"  # [enforce_avx512 == "absent" and blas_impl == "blis"]
```
Comment (mattip) [inline recipe/meta.yaml:71]: You can directly access the `np.core.multiarray._multiarray_umath.__cpu_features__['AVX512F']`, and we should think of a nicer way to expose this from NumPy.

Comment (h-vetinari) [reply recipe/meta.yaml:71]: Yeah, you're right, thanks (I now remember a similar discussion from https://github.com/numpy/numpy/issues/18058). But now that the test suite passes with blis both with/without avx512, I'll be able to drop those conditions again anyway.

**Comment by h-vetinari** [comment]
### Update for numpy 1.21.2: still all green! 🥳 

After 0 failure out of 76 for 1.21.1, we remained at 0 failures (out of 68; having removed the AVX512-specific runs for blis).

### Details

| lib | [before](https://github.com/conda-forge/numpy-feedstock/pull/237#issuecomment-894690400) | after | updated<br/>version | updated<br/>build |
|--------|-------|----|-|-|
| `numpy` | `1.21.1` | `1.21.2` | X | |
| `libblas` | `3.9.0-10` | `3.9.0-11` | | X |
| `blis` | `0.8.1-1` | `0.8.1-1` | | |
| `openblas` | `0.3.17-pthreads-1` | `0.3.17-pthreads-1` | | |
| `mkl` | `2021.3.0-557` | `2021.3.0-564` | | X |
| `netlib` | `3.9.0-5` | `3.9.0-5` | | |
| `pypy` | `7.3.5-7` | `7.3.5-9` | | X |

---------------

| variant| blis | mkl | netlib | openblas | sum* | 
|-----|-------|----|----|----|---|
| linux / x86 | ✔️ | ✔️ | ✔️ | ✔️ | - | 
| linux / aarch | ➖ | ➖ | ✔️ | ✔️ | - | 
| linux / ppc64le | ➖ | ➖ | ✔️ | ✔️ | - |
| osx / arm | ➖ | ➖ | ✔️ | ✔️ | - |
| osx / x86 | ✔️ | ✔️ | ✔️ | ✔️ | - |
| win / x86 | ✔️ | ✔️ | ✔️ | ✔️ | - |
| sum* | - | - | - | - | 0 |

`*` sum of Failures (out of a total of 68 CI combinations being tested)

Build logs:
[Azure](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=374722&view=results) 
[Drone](https://cloud.drone.io/conda-forge/numpy-feedstock/278)
[Travis](https://app.travis-ci.com/github/conda-forge/numpy-feedstock/builds/237394291)

**Comment by h-vetinari** [comment]
### Update for numpy 1.21.5: all green except PPC (as before)

Due to the missing `sys.exit` wrapper for `numpy.test`, we were missing some error reporting. In particular, the PPC builds were all failing before already, so with that in mind: After 10 failures (PPC-only) out of 68 runs for 1.21.3, we remained at 10 failures (PPC-only).

### Notable 
* One test that was skipped in #250 would be failing on osx (see https://github.com/numpy/numpy/issues/20637), but since it's part of the regular skips, I'm not counting it as a failure

### Details

| lib | [before](https://github.com/conda-forge/numpy-feedstock/pull/237#issuecomment-917245643) | after | updated<br/>version | updated<br/>build |
|--------|-------|----|-|-|
| `numpy` | `1.21.3` | `1.21.5` | X | |
| `libblas` | `3.9.0-12` | `3.9.0-12` | | |
| `blis` | `0.8.1-1` | `0.8.1-1` | | |
| `openblas` | `0.3.18-pthreads-0` | `0.3.18-pthreads-0` | | |
| `mkl` | `2021.4.0-729` | `2021.4.0-729` | | |
| `netlib` | `3.9.0-5` | `3.9.0-5` | | |
| `pypy` | `7.3.5-9` | `7.3.7-3` | X | |

---------------

| variant| blis | mkl | netlib | openblas | sum* | 
|-----|-------|----|----|----|---|
| linux / x86 | ✔️ | ✔️ | ✔️ | ✔️ | - | 
| linux / aarch | ➖ | ➖ | ✔️ | ✔️ | - | 
| linux / ppc64le | ➖ | ➖ | ✖️  | ✖️ | 10F  |
| osx / arm | ➖ | ➖ | ✔️ | ✔️ | - |
| osx / x86 | ✔️ | ✔️ | ✔️ | ✔️ | - |
| win / x86 | ✔️ | ✔️ | ✔️ | ✔️ | - |
| sum* | - | - | 5F | 5F | 10F |

`*` sum of Failures (out of a total of 68 CI combinations being tested)

Build logs:
[Azure](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=431370&view=results)

**Comment by mattip** [comment]
Nice. A heads-up that there is apparently a problem with the recently released OpenBLAS 0.3.19 and NumPy: see numpy/numpy#20660

**Comment by h-vetinari** [comment]
Not expecting a 1.21.6 release, so closing this.

**Comment by h-vetinari** [comment]
### Update for 1.21.6 (+ new PyPy builds and BLAS updates): all green except PPC (as before)

Turns out I guessed wrong about:
> Not expecting a 1.21.6 release, so closing this.

Also, due to the rebuilds for pypy3.8/3.9, much less several relevant BLAS (& infrastructure) changes, it makes sense to do an update here.

From 10 failures (PPC-only) out of 86 runs, we're now at _**12 failures (PPC-only) out of 108 runs**_.

### Notable

- Added `accelerate` BLAS flavour on osx
- Testing against PyPy 3.8 and 3.9 added everywhere but for osx-arm
- Version bumps for openblas, blis & MKL
- [Switched](https://github.com/conda-forge/numpy-feedstock/commit/8b869976b2c4da3f1f121ac66fd58a0486ed3878) to running the full test suite; emulation keeps running only `label='fast'` tests.

### Details

| variant | [before](https://github.com/conda-forge/numpy-feedstock/pull/237#issuecomment-1001939428) | after |
|--------|-------|----|
| linux + ppc | test failures due to emulation problems | as before |

--------------------

| lib | [before](https://github.com/conda-forge/numpy-feedstock/pull/237#issuecomment-1001939428) | after | updated<br/>version | updated<br/>build |
|--------|-------|----|-|-|
| `numpy` | `1.21.5` | `1.21.6` | X | |
| `libblas` | `3.9.0-12` | `3.9.0-14` | | X |
| `blis` | `0.8.1-1` | `0.9.0-0` | X | |
| `openblas` | `0.3.18-pthreads-1` | `0.3.20-pthreads-0` | X | |
| `mkl` | `2021.4.0-729` | `2022.0.1-803` | X | |
| `netlib` | `3.9.0-5` | `3.9.0-5` | | |
| `pypy` | ` 7.3.7-3` | `7.3.9-1` | X | |
| [`qemu-user-static`](https://github.com/multiarch/qemu-user-static) | ? | `6.1.0-8` | | |

---------------

| variant | accelerate | blis | mkl | netlib | openblas | sum* | 
|--------|------------|----|-----|-------|-----------|-------|
| linux / x86 | ➖ | ✔️ | ✔️ | ✔️ | ✔️ | - | 
| linux / aarch | ➖ | ➖ | ➖ | ✔️ | ✔️ | - | 
| linux / ppc64le | ➖ | ➖ | ➖ | ✖️ | ✖️ | 12F  |
| osx / arm | ✔️ | ➖ | ➖ | ✔️ | ✔️ | - |
| osx / x86 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | - |
| win / x86 | ➖ | ✔️ | ✔️ | ✔️ | ✔️ | - |
| sum* | - | - | - | 6F | 6F | 12F |

`*` sum of Failures (out of a total of 108 CI combinations being tested)

Build logs:
[Azure](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=491304&view=results)

**Pull request closed without merging**